book/src/linking_and_dividing.md

```diff
@@ -0,0 +1,241 @@
+# Linking and Dividing
+
+This section introduces everything thats needed to write JNI functions. Upon
+finishing this chapter, you'll have written a method that divides two integers
+using Rust.
+
+# Whats in NativeAPI?
+
+`NativeAPI` is a class that lives in `jnibookjava`, the Java side of the example
+code that accompanies this book. It uses `System.loadLibrary` to locate the
+`jnibookrs` lib, and specifies native methods using the `native` keyword. The
+native method called `verify_link` was used earlier to ensure that the project
+was setup correctly, with environment variables.
+
+For style reasons, we always add native methods to `NativeAPI`. It doesn't need
+to be that way, it's just a convention that we follow here.
+
+```java
+package com.github.jni_rs.jnibook;
+
+class NativeAPI {
+
+   private static final Throwable INIT_ERROR;
+
+   // The static block will be executed the first time the NativeAPI
+   // class is used.
+   static {
+       Throwable error = null;
+       try {
+           System.loadLibrary("jnibookrs");
+       } catch (Throwable t) {
+           error = t;
+       }
+       INIT_ERROR = error;
+   }
+
+   private NativeAPI() {
+       // Not instantiable
+   }
+
+   static native int verify_link();
+}
+```
+
+## Function Linking Rules
+To get a native Java function to correctly bind to a Rust function underneath,
+we need to ensure that it's named properly, and that the parameters and return
+type match. First, we'll discuss getting the Rust function name.
+
+### Naming the Rust Function
+JNI has a long list of rules that it uses to encode the package path, class name
+and function name into a valid C function name. Fortunately, you don't have to
+learn them all, as `javac` can be used to generate a C function for you, which
+you can then copy for use with Rust. For this example, we'll work with
+`NativeAPI` and the `verify_link` function.
+
+To get a name for a function, follow these steps.
+
+1. Add a `native` function to a Java class. Remember that the package path,
+   class name, and function name are all encoded into the C function name, so
+   make sure those are as you want them. 
+2. Run `javac -h . NativeAPI.java` to produce C headers. (We assume that the
+   class resides in `NativeAPI.java`, but it doesn't have to.)
+3. Copy the C function name out of the header file it produced.
+
+Upon completing these steps, you will see a file called
+`com_github_jni_rs_jnibook_NativeAPI.h` that contains the following:
```

Comment:
Nit: We probably shall not use com.github namespace. For a tutorial Java code that isn't uploaded on Central, we may use a shorter name, like jni_rs.jnibook or just jnibook.

Reply:
Sounds good to me. Maybe jnirsbook?

Reply:
👍 , or "jni_rs_book" so that it's slightly more readable.